internal/proxy/proxy.go

```diff
@@ -0,0 +1,84 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/buzzfeed/sso/internal/pkg/hostmux"
+	"github.com/buzzfeed/sso/internal/pkg/options"
+	"github.com/buzzfeed/sso/internal/pkg/ping"
+)
+
+type SSOProxy struct {
+	http.Handler
+}
+
+func New(opts *Options) (*SSOProxy, error) {
+	optFuncs := []func(*OAuthProxy) error{}
+
+	var requestSigner *RequestSigner
+	var err error
+	if opts.RequestSigningKey != "" {
+		requestSigner, err = NewRequestSigner(opts.RequestSigningKey)
+		if err != nil {
+			return nil, err
+		}
+		optFuncs = append(optFuncs, SetRequestSigner(requestSigner))
+	}
+
+	if len(opts.EmailAddresses) != 0 {
+		optFuncs = append(optFuncs, SetValidator(options.NewEmailAddressValidator(opts.EmailAddresses)))
+	} else {
+		optFuncs = append(optFuncs, SetValidator(options.NewEmailDomainValidator(opts.EmailDomains)))
+	}
+
+	hostRouter := hostmux.NewRouter()
+	for _, upstreamConfig := range opts.upstreamConfigs {
+		var handler http.Handler
+		switch route := upstreamConfig.Route.(type) {
+		case *SimpleRoute:
+			reverseProxy := NewReverseProxy(route.ToURL, upstreamConfig)
+			handler, _ = NewReverseProxyHandler(
```

Comment:
It's not an error, but a set of `tags` that looks like we're never wired up to anything super useful. I resisted changing the function signature since I rewrite the reverse proxy handling/handler in https://github.com/buzzfeed/sso/pull/189